include/circt/Dialect/SSP/Utilities.h

```diff
@@ -0,0 +1,437 @@
+  auto graphOp = instOp.getDependenceGraph();
+  graphOp.walk([&](OperationOp opOp) {
+    prob.insertOperation(opOp);
+    loadOperationProperties<ProblemT, OperationPropertyTs...>(
+        prob, opOp, opOp.getPropertiesAttr());
+  });
+
+  // Then walk them again, and load auxiliary dependences as well as any
+  // dependence properties.
+  graphOp.walk([&](OperationOp opOp) {
+    ArrayAttr depsAttr = opOp.getDependencesAttr();
+    if (!depsAttr)
+      return;
+
+    for (auto depAttr : depsAttr.getAsRange<DependenceAttr>()) {
+      Dependence dep;
+      if (FlatSymbolRefAttr sourceRef = depAttr.getSourceRef()) {
+        Operation *sourceOp = SymbolTable::lookupSymbolIn(graphOp, sourceRef);
+        assert(sourceOp);
+        dep = Dependence(sourceOp, opOp);
+        LogicalResult res = prob.insertDependence(dep);
+        assert(succeeded(res));
+        (void)res;
+      } else
+        dep = Dependence(&opOp->getOpOperand(depAttr.getOperandIdx()));
+
+      loadDependenceProperties<ProblemT, DependencePropertyTs...>(
+          prob, dep, depAttr.getProperties());
+    }
+  });
+
+  return prob;
+}
+
+//===----------------------------------------------------------------------===//
+// circt::scheduling::Problem (or subclasses) -> ssp.InstanceOp
+//===----------------------------------------------------------------------===//
+
+template <typename ProblemT, typename... OperationPropertyTs>
+ArrayAttr saveOperationProperties(ProblemT &prob, Operation *op,
+                                  ImplicitLocOpBuilder &b) {
+  SmallVector<Attribute> props;
+  Attribute prop;
+  ((prop = OperationPropertyTs::getFromProblem(prob, op, b.getContext()),
+    prop ? props.push_back(prop) : (void)prop),
+   ...);
+  return props.empty() ? ArrayAttr() : b.getArrayAttr(props);
+}
+
+template <typename ProblemT, typename... OperatorTypePropertyTs>
+ArrayAttr saveOperatorTypeProperties(ProblemT &prob, OperatorType opr,
+                                     ImplicitLocOpBuilder &b) {
+  SmallVector<Attribute> props;
+  Attribute prop;
+  ((prop = OperatorTypePropertyTs::getFromProblem(prob, opr, b.getContext()),
+    prop ? props.push_back(prop) : (void)prop),
+   ...);
+  return props.empty() ? ArrayAttr() : b.getArrayAttr(props);
+}
+
+template <typename ProblemT, typename... DependencePropertyTs>
+ArrayAttr saveDependenceProperties(ProblemT &prob, Dependence dep,
+                                   ImplicitLocOpBuilder &b) {
+  SmallVector<Attribute> props;
+  Attribute prop;
+  ((prop = DependencePropertyTs::getFromProblem(prob, dep, b.getContext()),
+    prop ? props.push_back(prop) : (void)prop),
+   ...);
+  return props.empty() ? ArrayAttr() : b.getArrayAttr(props);
+}
+
+template <typename ProblemT, typename... InstancePropertyTs>
+ArrayAttr saveInstanceProperties(ProblemT &prob, ImplicitLocOpBuilder &b) {
+  SmallVector<Attribute> props;
+  Attribute prop;
+  ((prop = InstancePropertyTs::getFromProblem(prob, b.getContext()),
```

Comment:
Is this form required? It may just be me, but coming from the LLVM coding standards, it seems a little dense and surprising. I've never seen the `...` in LLVM/MLIR C++ code like this.

Reply:
> Is this form required?

No, in this case, the alternative would be to have helper templates that unpack the type parameters by recursion.

> It may just be me, but coming from the LLVM coding standards, it seems a little dense and surprising.

Funnily, I had the same reaction when I first learned about the fold expressions from @7FM 🤓 However, I find the construct quite neat now. I added comments to point out what's happening.

> I've never seen the ... in LLVM/MLIR C++ code like this.

Fold expressions are a C++17 feature, but [people would like to use them in the future](https://github.com/llvm/llvm-project/blob/4cf0a9d4ae76f46ee2833f8479bb712b072673f3/mlir/include/mlir/IR/PatternMatch.h#L1417).

Reply:
Got it, if this is a new, well understood c++17 feature, then sounds good to me. Glad we have the ability to use the "new" stuff in CIRCT.